### PR TITLE
[NOSQUASH] refacto: don't permit to access to RenderingEngine::filesystem widely

### PR DESCRIPTION
**Pull request by nerzhul**
This PR include multiple changes:

* 2 old irrlicht version compat removal (1.7 and less)
* Don't permit to access to Rendering::filesystem pointer from everywhere. It will permit better code, and prevent calling from bad thread if we need to change the threading model on client (and i want to change it on the client part, to be more event based)
* Move extractZipFile logic bound directly in the main menu lua to the client. It's not lua stack role to perform all this logic. This also permit to have a re-usable function
* hide mesh cache operations inside RenderingEngine, it's not exposed anymore
* No more access to the singleton instance from everywhere (RenderingEngine::get_instance dropped)
* RenderingEngine::get_timer_time is now non static
* RenderingEngine::draw_menu_scene is now non static
* RenderingEngine::draw_scene is now non static
* RenderingEngine::{initialize,finalize} are now non static
* RenderingEngine::run is now non static
* RenderingEngine::getWindowSize now have a static helper. It was mandatory to hide the global get_instance access
* RenderingEngine::get_scene_manager  is now non static
* Rename Game::sky to Game::m_sky
* don't use RenderingEngine singleton on CAO, instead call the proper children we need
* drop unused Hud::smgr & Hud::guienv
* Hud: make driver, client, player, inventory, tsrc private
* Fix wrong hud.h called on c_content.h

## To do

This PR is Ready for Review.

I hope it can be  review soon to prevent a huge list of changed, but they are interdependent.

Don't squash commits. It's non function changes but if we need to bisect it will make the work easier

### CODE REVIEW

File: src/client/client.cpp [hunk 1]
```diff
@@ -735,6 +725,72 @@ bool Client::loadMedia(const std::string &data, const std::string &filename,
...
+	/**********************************************************************/
+	/* WARNING this is not threadsafe!!                                   */
+	/**********************************************************************/
+	io::IFileArchive* opened_zip = fs->getFileArchive(fs->getFileArchiveCount() - 1);
+
+	const io::IFileList* files_in_zip = opened_zip->getFileList();
+
+	unsigned int number_of_files = files_in_zip->getFileCount();
+
+	for (unsigned int i=0; i < number_of_files; i++) {
```
Comment (JosiahWI) [inline src/client/client.cpp:747]: Could this infinite loop? I also suggest ++i instead of i++ here but that's a very minor performance point.

Comment (nerzhul) [reply src/client/client.cpp:747]: it's very minor and only called when loading media on connection

File: src/client/game.cpp [hunk 2]
```diff
@@ -4063,27 +4065,6 @@ void Game::readSettings()
  ****************************************************************************/
 /****************************************************************************/
 
-void Game::extendedResourceCleanup()
```
Comment (nerzhul) [inline src/client/game.cpp:4066]: note: this function is pretty useless, it's a 7 years old function to do some debug

**Review by SmallJoker — APPROVED** [review]
OK. Minetest still works; code is according to the style guidelines.

File: src/nodedef.cpp [hunk 3]
```diff
@@ -1456,7 +1454,7 @@ void NodeDefManager::updateTextures(IGameDef *gamedef,
 	for (u32 i = 0; i < size; i++) {
 		ContentFeatures *f = &(m_content_features[i]);
 		f->updateTextures(tsrc, shdsrc, meshmanip, client, tsettings);
-		progress_callback(progress_callback_args, i, size);
+		((Client *)gamedef)->showUpdateProgressTexture(progress_callback_args, i, size);
```
Comment (sfan5) [inline src/nodedef.cpp:1457]: `client` exists in this function with the correct type

File: src/script/common/c_content.h [hunk 4]
```diff
@@ -39,7 +39,7 @@ extern "C" {
 #include "itemgroup.h"
 #include "itemdef.h"
 #include "c_types.h"
-#include "hud.h"
+#include "../../hud.h"
```
Comment (sfan5) [inline src/script/common/c_content.h:42]: was this moved or why is this necessary?

**Comment by nerzhul** [comment]
@sfan5 RenderingEngine::get_instance() has been removed. It's not possible to access to the render everywhere in the code.
Idea is to ensure the renderer is only called from legit objects. Currently our client is a crap of spaghettis, and adding a thread for anything will just make it a pure hell.
I want to reduce the scope of the renderer exposition, then the client object (which is like a global var as it's stored nearly everywhere in the client for arious reasons)
My idea on the long term is to make the client network part event based and not mono threaded. this will permit to stop freeze the client when you receive large amount of packets to proceed. This will also permit to offload some huge processing before rendering outside of this main thread (ie. decompressing mapblocks)

File: src/nodedef.cpp (hunk 3 above)
Comment (nerzhul) [reply src/nodedef.cpp]: fixed, my bad :)
